### PR TITLE
Feature/reminders with firebase data source

### PR DESCRIPTION
**Pull request by StanislavPrime**
Add firebase as remote data source for the create, read, update and delete operations of the dashboard and reminder list

### CODE REVIEW

File: examples/reminders/lib/base/di/app_dependencies.dart [hunk 1]
```diff
@@ -63,14 +65,14 @@ class AppDependencies {
...
 
-  List<Provider> get _analytics => [
-        Provider<FirebaseAnalytics>(
-            create: (context) => FirebaseAnalytics.instance),
-        Provider<FirebaseAnalyticsObserver>(
-          create: (context) =>
-              FirebaseAnalyticsObserver(analytics: context.read()),
-        ),
-      ];
+  // List<Provider> get _analytics => [
```
Comment (StanevPrime) [inline examples/reminders/lib/base/di/app_dependencies.dart:68]: Please remove the commented lines if they are not neede

File: examples/reminders/lib/base/di/app_dependencies.dart [hunk 2]
```diff
@@ -6,8 +6,9 @@
 // https://opensource.org/licenses/MIT.
 
 import 'package:dio/dio.dart';
-import 'package:firebase_analytics/firebase_analytics.dart';
-import 'package:firebase_messaging/firebase_messaging.dart';
+// import 'package:firebase_analytics/firebase_analytics.dart';
```
Comment (StanevPrime) [inline examples/reminders/lib/base/di/app_dependencies.dart:9]: Same as below

File: examples/reminders/lib/base/data_sources/remote/reminders_remote_data_source_factory.dart [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import '../../app/config/environment_config.dart';
+import '../local/reminders_local_data_source.dart';
+import 'reminders_data_source.dart';
+import 'reminders_firebase_data_source.dart';
+
+class RemindersRemoteDataSourceFactory {
+  static RemindersDataSource withFirebaseDataSource() =>
```
Comment (StanevPrime) [inline examples/reminders/lib/base/data_sources/remote/reminders_remote_data_source_factory.dart:7]: Please use the encapsulation principle, in other words hide (make private) all methods that should not be used direcly

File: examples/reminders/lib/base/app/reminders.dart [hunk 4]
```diff
@@ -55,37 +55,37 @@ class __MyMaterialAppState extends State<_MyMaterialApp> {
...
-
-  void _addInterceptors(){
-    context.read<Dio>().interceptors.addAll([
-      AuthInterceptor(context.read(), context.read(), context.read()),
-      AnalyticsInterceptor(context.read()),
-
-      /// TODO: Add your own interceptors here
-    ]);
-  }
+  // Future<void> _configureFCM() async {
```
Comment (StanevPrime) [inline examples/reminders/lib/base/app/reminders.dart:64]: Remove all commented code if not needed

File: examples/reminders/lib/base/data_sources/remote/reminders_firebase_data_source.dart [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+      dueDate: dueDate,
+      id: remindersReference.doc().id,
+      title: title,
+      complete: complete,
+    );
+
+    var createdReminder = await remindersReference.add(reminder.toJson());
+    var createdReminderId = createdReminder.id;
+    reminder = reminder.copyWith(id: createdReminderId);
+    await remindersReference.doc(createdReminderId).update(reminder.toJson());
```
Comment (StanevPrime) [inline examples/reminders/lib/base/data_sources/remote/reminders_firebase_data_source.dart:37]: Hmm what is the reason of updating the document right after you have created it ?

File: examples/reminders/lib/base/data_sources/remote/reminders_firebase_data_source.dart [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    // Get the result of the query
+    final snap = await querySnapshot.get();
+    if (request.pageSize == 10) {
+      lastFetchedRecord = snap.docs.last;
+    }
+
+    final reminders = snap.docs.asReminderList();
+
+    if (request.page == 1) {
+      final length = await remindersReference.get();
```
Comment (StanevPrime) [inline examples/reminders/lib/base/data_sources/remote/reminders_firebase_data_source.dart:67]: Yep, this is really expensive operation

File: examples/reminders/lib/base/models/reminder/reminder_model.dart [hunk 7]
```diff
@@ -1,3 +1,4 @@
+import 'package:cloud_firestore/cloud_firestore.dart';
```
[the remark this answers is not in the document]
Comment (StanislavPrime) [reply examples/reminders/lib/base/models/reminder/reminder_model.dart:1]: I was using `as Timestamp` from it, but I have removed it

File: examples/reminders/lib/base/data_sources/remote/reminders_remote_data_source_factory.dart (hunk 3 above)
Comment (StanislavPrime) [reply examples/reminders/lib/base/data_sources/remote/reminders_remote_data_source_factory.dart:7]: I have made private the two methods

File: examples/reminders/lib/base/di/app_dependencies.dart (hunk 1 above)
Comment (StanislavPrime) [reply examples/reminders/lib/base/di/app_dependencies.dart]: The comment lines were removed and a check was added for the `EnvironmentConfig` value

File: examples/reminders/lib/base/di/app_dependencies.dart (hunk 2 above)
Comment (StanislavPrime) [reply examples/reminders/lib/base/di/app_dependencies.dart]: The comment lines were uncommented and a check `config != EnvironmentConfig.dev` was added

File: examples/reminders/lib/base/data_sources/remote/reminders_firebase_data_source.dart (hunk 6 above)
Comment (StanislavPrime) [reply examples/reminders/lib/base/data_sources/remote/reminders_firebase_data_source.dart]: I have replaced the fetching of the elements with fetching a counter value stored in firebase

**Review by StanevPrime — APPROVED** [review]

**Pull request merged**